### ISSUE TITLE
TypeError: unsupported operand type(s) for |: 'ABCMeta' and 'NoneType'

### ISSUE THREAD

**Issue by joshooaj** [squidfunk/mkdocs-material#6264]
### Bug description

When using the latest available version of the mkdocs-material-insiders theme in a debian bullseye devcontainer, I cannot build/serve/publish with the blog plugin enabled without getting the following error.

It appears similar to #5881 in that my environment has a <3.10 version of python (3.9.2) and the info plugin cannot package the reproduction while the blog plugin is enabled. So heads up - in the attached reproduction I have commented out the blog plugin.

```plaintext
PS /workspaces/mkdocs-bug> mkdocs build
Traceback (most recent call last):
  File "/usr/local/bin/mkdocs", line 8, in <module>
    sys.exit(cli())
  File "/usr/lib/python3/dist-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3/dist-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3/dist-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3/dist-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/mkdocs/__main__.py", line 283, in build_command
    cfg = config.load_config(**kwargs)
  File "/usr/local/lib/python3.9/dist-packages/mkdocs/config/base.py", line 378, in load_config
    errors, warnings = cfg.validate()
  File "/usr/local/lib/python3.9/dist-packages/mkdocs/config/base.py", line 230, in validate
    run_failed, run_warnings = self._validate()
  File "/usr/local/lib/python3.9/dist-packages/mkdocs/config/base.py", line 188, in _validate
    self[key] = config_option.validate(value)
  File "/usr/local/lib/python3.9/dist-packages/mkdocs/config/config_options.py", line 182, in validate
    return self.run_validation(value)
  File "/usr/local/lib/python3.9/dist-packages/mkdocs/config/config_options.py", line 1064, in run_validation
    self.load_plugin_with_namespace(name, cfg)
  File "/usr/local/lib/python3.9/dist-packages/mkdocs/config/config_options.py", line 1102, in load_plugin_with_namespace
    return (name, self.load_plugin(name, config))
  File "/usr/local/lib/python3.9/dist-packages/mkdocs/config/config_options.py", line 1120, in load_plugin
    plugin_cls = self.installed_plugins[name].load()
  File "/usr/local/lib/python3.9/dist-packages/importlib_metadata/__init__.py", line 209, in load
    module = import_module(match.group('module'))
  File "/usr/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 790, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/usr/local/lib/python3.9/dist-packages/material/plugins/blog/plugin.py", line 50, in <module>
    from .structure import Archive, Category, Excerpt, Reference, Post, View
  File "/usr/local/lib/python3.9/dist-packages/material/plugins/blog/structure/__init__.py", line 42, in <module>
    from .markdown import ExcerptTreeprocessor
  File "/usr/local/lib/python3.9/dist-packages/material/plugins/blog/structure/markdown.py", line 31, in <module>
    class ExcerptTreeprocessor(Treeprocessor):
  File "/usr/local/lib/python3.9/dist-packages/material/plugins/blog/structure/markdown.py", line 34, in ExcerptTreeprocessor
    def __init__(self, page: Page, base: Page | None = None):
TypeError: unsupported operand type(s) for |: 'ABCMeta' and 'NoneType'
```

### Related links

- [Reporting a bug](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/)

### Reproduction

[9.4.7+insiders.4.42.3-blog-plugin-typeerror.zip](https://github.com/squidfunk/mkdocs-material/files/13221032/9.4.7%2Binsiders.4.42.3-blog-plugin-typeerror.zip)

### Steps to reproduce

1. Use the 9.4.7+insiders.4.42.3 release of mkdocs-material-insiders.
   - I'm specifically installing using the syntax `pip install git+https://github.com/squidfunk/mkdocs-material-insiders.git` but it looks like I also get the same error when using the build associated with the 9.4.7+insiders.4.42.3 tag connected to the published release.
3. Include the blog plugin in the list of plugins in `mkdocs.yml`
4. Optional: Scaffold the blog folder and first post. In testing, it ended up not mattering if the `blog` folder was present under `docs/` as the error occurs on `__init__`.
5. Run `mkdocs build`

### Before submitting

- [X] I have read and followed the [bug reporting guidelines](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/).
- [X] I have attached links to [the documentation](https://squidfunk.github.io/mkdocs-material/), and possibly related [issues](https://github.com/squidfunk/mkdocs-material/issues) and [discussions](https://github.com/squidfunk/mkdocs-material/discussions).
- [X] I assure that I have [removed all customizations](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/#remove-customizations) before submitting this bug report.
- [X] I have attached a __.zip file__ with a [minimal reproduction](https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/).

**Comment by alexvoss** [comment]
I noticed that there are two lines for Jinja in `requirements.lock.txt` - as well as for other packages. Could this be part of the problem? Could you try again with a fresh venv?

**Comment by joshooaj** [comment]
Thanks @alexvoss,

My environment is a new container created from the base image [mcr.microsoft.com/powershell:lts-debian-11](https://github.com/PowerShell/PowerShell-Docker/blob/master/release/7-2/debian11/docker/Dockerfile) and prior to running `mkdocs build` I ran the following commands:

```sh
apt update && apt install git python3-pip mkdocs
pip install git+https://github.com/squidfunk/mkdocs-material-insiders.git
```

It looks like the duplicate entries in `requirements.lock.txt` are a result of having installed mkdocs via `apt install` before installing `mkdocs-material` with `pip install git+https://github.com/squidfunk/mkdocs-material-insiders.git`.

It doesn't appear like this is the issue though - I made another log package where the container was prepared with the following commands:

```sh
apt update && apt install git python3-pip
pip install git+https://github.com/squidfunk/mkdocs-material-insiders.git
```

[9.4.7+insiders.4.42.3-blog-plugin-typerror.zip](https://github.com/squidfunk/mkdocs-material/files/13229810/9.4.7%2Binsiders.4.42.3-blog-plugin-typerror.zip)

There don't appear to be any duplicate entries in `requirements.lock.txt` anymore:

| Package                    | Version               |
|----------------------------|-----------------------|
| Babel                      | 2.13.1                |
| certifi                    | 2023.7.22             |
| charset-normalizer         | 3.3.2                 |
| click                      | 8.1.7                 |
| colorama                   | 0.4.6                 |
| ghp-import                 | 2.1.0                 |
| idna                       | 3.4                   |
| importlib-metadata         | 6.8.0                 |
| Jinja2                     | 3.1.2                 |
| Markdown                   | 3.5.1                 |
| MarkupSafe                 | 2.1.3                 |
| mergedeep                  | 1.3.4                 |
| mkdocs                     | 1.5.3                 |
| mkdocs-material            | 9.4.7+insiders.4.42.3 |
| mkdocs-material-extensions | 1.3                   |
| packaging                  | 23.2                  |
| paginate                   | 0.5.6                 |
| pathspec                   | 0.11.2                |
| pip                        | 20.3.4                |
| platformdirs               | 3.11.0                |
| Pygments                   | 2.16.1                |
| pymdown-extensions         | 10.3.1                |
| python-dateutil            | 2.8.2                 |
| PyYAML                     | 6.0.1                 |
| pyyaml_env_tag             | 0.1                   |
| regex                      | 2023.10.3             |
| requests                   | 2.31.0                |
| setuptools                 | 52.0.0                |
| six                        | 1.16.0                |
| urllib3                    | 2.0.7                 |
| watchdog                   | 3.0.0                 |
| wheel                      | 0.34.2                |
| zipp                       | 3.17.0                |

**Comment by alexvoss** [comment]
Thanks for the update. I have installed Python 3.9.16 (via anaconda/micromamba) and then Insiders:

```
micromamba create --name python3.9 -c anaconda python=3.9.16
pip install git+https://${GH_TOKEN}@github.com/squidfunk/mkdocs-material-insiders.git@9.4.7-insiders-4.42.3
```

This way I was able to reproduce the problem on a Mac.

**Comment by squidfunk** [comment]
I'm pretty sure we're missing `from __future__ import annotations` here.

**Comment by alexvoss** [comment]
Thanks, @squidfunk, I will wire together an environment to test your suggestion. Really should go about automating that...

**Comment by alexvoss** [comment]
Might be an idea to hunt for other occurrences of this issue? Is this line something that can/should be added to all Python files? I really need to learn more about type annotations in Python...

**Comment by joshooaj** [comment]
Thanks guys!

I did a quick regex against all the .py files in the insiders repo for what I _think_ catches the annotation pattern "Page | None = None", `\b\w+\s*\|\s*\w+\s*=\s*\w+`, and the only other file that came up was `material/plugins/privacy/plugin.py` which already has the `from __future__ import annotations` import.

**Comment by alexvoss** [comment]
Thanks, that is good to know.

**Comment by squidfunk** [comment]
We definitely need to set up some smoke/regression tests. I started working on this, but currently have other priorities. Those problems will then be catched before releases ☺️

**Comment by squidfunk** [comment]
Fixed in [`e7a6a23`](https://github.com/squidfunk/mkdocs-material-insiders/commit/e7a6a238e163b86b435dcb77d5b706b3e453ab7a).

**Comment by alexvoss** [comment]
> We definitely need to set up some smoke/regression tests. I started working on this, but currently have other priorities. Those problems will then be catched before releases ☺️

Would this involve setting up another job in the GitHub Actions workflow with a different setting for `PYTHON_VERSION`? If so then I could put that on my plate. I might just rise to the challenge and learn about GitHub Actions, which I have not used before - more of a Jenkins type, me ;o) Does that show how old I am?

**Comment by squidfunk** [comment]
We could leverage build matrixes – [check out MkDocs' workflow](https://github.com/mkdocs/mkdocs/blob/master/.github/workflows/ci.yml), they do it for different Python versions. For this, we first need tests, though 😉 I've worked with Jenkins, but IMHO, GitHub Actions integrates so well with GitHub, and is free for public repositories, it's like a no-brainer + you can share workflows with other users.

**Comment by alexvoss** [comment]
Yes, that looks good. I may be able to look at this tonight. Was not questioning the use of GitHub Actions, just wanted to say that I am familiar with the concept of CI/CD but just not what Actions. You did suggest first doing smoke tests, which should be possible.

If I find opportunities to add unit tests, what framework should I use? [pytest](https://docs.pytest.org/en/latest/)? I am asking because writing unit tests might be a good way for me to dig into how some plugins work in Material. I have not yet found the time to study the implementations of blog or project but might have a go over the WE.

**Comment by squidfunk** [comment]
I started with unittest and was quite happy with it.

**Comment by squidfunk** [comment]
Released as part of [9.4.8+insiders-4.43.0](https://squidfunk.github.io/mkdocs-material/insiders/changelog/#4.43.0).
